torch/optim/adamax.py

```diff
@@ -107,15 +111,120 @@ def step(self, closure=None):
-                     exp_infs,
-                     state_steps,
-                     eps=eps,
-                     beta1=beta1,
-                     beta2=beta2,
-                     lr=lr,
-                     weight_decay=weight_decay)
+            adamax(params_with_grad,
+                   grads,
+                   exp_avgs,
+                   exp_infs,
+                   state_steps,
+                   foreach=foreach,
+                   eps=eps,
+                   beta1=beta1,
+                   beta2=beta2,
+                   lr=lr,
+                   weight_decay=weight_decay)
 
         return loss
+
+
+def adamax(params: List[Tensor],
+           grads: List[Tensor],
+           exp_avgs: List[Tensor],
+           exp_infs: List[Tensor],
+           state_steps: List[int],
+           foreach: bool = False,
+           *,
+           eps: float,
+           beta1: float,
+           beta2: float,
+           lr: float,
+           weight_decay: float):
+    r"""Functional API that performs adamax algorithm computation.
+
+    See :class:`~torch.optim.Adamax` for details.
+    """
+
+    if foreach and not torch.jit.is_scripting():
+        func = _multi_tensor_adamax
+    else:
+        func = _single_tensor_adamax
+
+    func(params,
+         grads,
+         exp_avgs,
+         exp_infs,
+         state_steps,
+         eps=eps,
+         beta1=beta1,
+         beta2=beta2,
+         lr=lr,
+         weight_decay=weight_decay)
+
+
+def _single_tensor_adamax(params: List[Tensor],
+                          grads: List[Tensor],
+                          exp_avgs: List[Tensor],
+                          exp_infs: List[Tensor],
+                          state_steps: List[int],
+                          *,
+                          eps: float,
+                          beta1: float,
+                          beta2: float,
+                          lr: float,
+                          weight_decay: float):
+
+    for i, param in enumerate(params):
+        grad = grads[i]
+        exp_avg = exp_avgs[i]
+        exp_inf = exp_infs[i]
+        step = state_steps[i]
+
+        if weight_decay != 0:
+            grad = grad.add(param, alpha=weight_decay)
+
+        # Update biased first moment estimate.
+        exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
+        # Update the exponentially weighted infinity norm.
+        norm_buf = torch.cat([
+            exp_inf.mul_(beta2).unsqueeze(0),
+            grad.abs().add_(eps).unsqueeze_(0)
+        ], 0)
+        torch.amax(norm_buf, 0, keepdim=False, out=exp_inf)
+
+        bias_correction = 1 - beta1 ** step
+        clr = lr / bias_correction
+
+        param.addcdiv_(exp_avg, exp_inf, value=-clr)
+
+
+def _multi_tensor_adamax(params: List[Tensor],
+                         grads: List[Tensor],
+                         exp_avgs: List[Tensor],
+                         exp_infs: List[Tensor],
+                         state_steps: List[int],
```

Comment:
The original (adamax)[https://github.com/pytorch/pytorch/blob/master/torch/optim/_multi_tensor/_functional.py#L73] in `optim/_multi_tensor/functional.py` takes in `state: List[Dict]` instead, but `states` was not used other than to access `state['steps']` for each `state` in `states` so I changed this to take in `state_steps` so the inputs would be the same as the single tensor version, is this ok?

Reply:
I don't think we have to take exactly the same thing since these are not public and we have a function above that can massage the arguments. But it is nicer if we can for sure.